ruby/phase-1.md

```diff
@@ -0,0 +1,434 @@
+can choose to gradually start overriding the defaults until they achieve what feels comfortable.*
+
+
+## Client
+
+The client is the primary point for interacting with workflows and async activities as well as
+managing namespaces. As hinted in the previous section there's a global singleton client available
+for applications that only connect to a single Temporal server and don't need varying
+configurations:
+
+```ruby
+# Global instance of the Client configured via `Temporal.configure`
+Temporal.start_workflow(MyWorkflow)
+
+# Explicitly configured local instance of the Client
+config = Temporal::Configuration.new(url: 'localhost:7233')
+client = Temporal::Client.new(config)
+client.start_workflow(MyWorkflow)
+```
+
+Despite the differences both clients offer exactly the same interface (the singleton client is an
+instance of a regular client stored on Temporal module for convenience). The interface will look
+something like this:
+
+```ruby
+class Client
+  def initialize(config)
+
+  def start_workflow(
+    workflow, # workflow class or a name
+    *args, # any args to be passed as the input to the workflow
+    options: {}, # execution options overrides
+    **kwargs # keyword arguments as inputs are also supported
+  ) -> WorkflowHandle
+
+  def workflow_handle(namespace, workflow_id, run_id = nil) -> WorkflowHandle
+end
+```
+
+Notes:
+
+- This is not a complete definition of the client
+- Both named and keyword arguments are supported, except for the reserved `options:`
+
+The combination of the singleton client and implicit execution attributes allows us to optionally
+support direct workflow invocations (removing all SDK traces out of the way):
+
+```ruby
+MyWorkflow.start(arg_1, arg_2, keyword_arg: arg_3)
+```
+
+Once a workflow is started all the interactions will be going through the handle object returned to
+the caller. The handle can of course be created from a namespace/workflow_id/run_id combination
+using the `Client#workflow_handle` method. Here the interface of the handle:
+
+```ruby
+class WorkflowHandle
+  def initialize(connection, namespace, workflow_id, run_id = nil)
+
+  def result(await: false, timeout: nil) -> Any
+
+  def describe -> WorkflowExecutionInfo
+
+  def cancel -> void
+
+  def query(name, *args) -> Any
+
+  def signal(name, *args) -> void
+
+  def terminate(reason, details: nil) -> void
+end
+
+```
+
+
+## Worker
+
+The worker interface allows SDK users to register any number of both workflows and activities. These
+will then be polled for on the specified namespaces & task queues and executed. The worker can poll
+on multiple task queues at the same time. These will be determined by the registered activities and
+workflow. If there are any overlaps we'll only create one poller for each unique task queue.
+
+```ruby
+config Temporal::Configuration.new(url: 'localhost:7233')
+worker = Temporal::Worker.new(config)
+
+worker.register_workflow(NewWorkflow)
```

Comment:
I would discourage delayed registration of activities and workflows on the worker and instead only accepts those during construction to prevent registration on a running worker.

Reply:
This can be controlled from within `register_*` method (if worker has been started — raise), but I wonder why does that have to be a limitation?

Reply:
We would like in the future to add worker queries where a worker can list which workflows and activities it has registered.
Even if we disallowed late registration, using these methods will cause a runtime error, this would be eliminated if registration is limited to construction time.